src/sa.py

```diff
@@ -1,15 +1,277 @@
+    elif fl == "first":
+        for i in range(len(bucket)):
+            if bucket[i] == -1:
+                return i
+
+def first_pass(buckets, bucket_out, LMS_array, LMS_blocks):
+    for i in range(len(LMS_array) -1, -1, -1):
+        index = first_last_index(bucket_out[LMS_blocks[i][0]], "last")
+        bucket_out[LMS_blocks[i][0]][index] = LMS_array[i]
+    return bucket_out
+
+def second_pass(bucket_out, LS_array, string):
+    for key in bucket_out:
+        for element in bucket_out[key]:
+            if element < 1:
+                continue
+            if LS_array[element - 1] == "L":
+                index = first_last_index(bucket_out[string[element - 1]], "first")
+                bucket_out[string[element - 1]][index] = element - 1
+    return bucket_out
+
+def bucket_indices(bucket_out):
+    indices = {}
+    for key in bucket_out:
+        indices[key] = len(bucket_out[key]) - 1
+    return indices
+
+def third_pass(bucket_out, LS_array, string):
+    indices = bucket_indices(bucket_out)
+
+    for key in sorted(bucket_out, reverse=True):
+        for i in range(len(bucket_out[key]) - 1, -1, -1):
+            if bucket_out[key][i] < 1:
+                continue
+            elif LS_array[bucket_out[key][i] - 1] == "S":
+                buck = string[bucket_out[key][i] - 1]
+                bucket_out[buck][indices[buck]] = bucket_out[key][i] - 1
+                indices[buck] -= 1
+    return bucket_out
+
+def number_LMS(bucket_out, LMS_blocks, LMS_array):
+    reduced = [0 for i in range(len(LMS_array))]
+    previous = None
+    count = -1
+    for key in bucket_out:
+        for element in bucket_out[key]:
+            if element in LMS_array:
+                block = LMS_blocks[LMS_array.index(element)]
+                if block != previous:
+                    count += 1
+                previous = block
+                reduced[LMS_array.index(element)] = str(count)
+    return "".join(reduced)
+
+def compute_t1(string):
+    LS_array = annotate_string(string)
+    LMS_array = get_LMS(LS_array)
+    LMS_blocks = get_LMS_blocks(string, LMS_array)
+    buckets = impute_buckets(string)
+    bucket_out = output_bucket(buckets)
+    bucket_out = first_pass(buckets, bucket_out, LMS_array, LMS_blocks)
+    bucket_out = second_pass(bucket_out, LS_array, string)
+    bucket_out = third_pass(bucket_out, LS_array, string)
+    reduced = 0#number_LMS(bucket_out, LMS_blocks, LMS_array)
+    return reduced, bucket_out
+
+#def SAIS(string):
+#    stack = list()
+#    while True:
+#        reduced, bucket_out = compute_t1
+#        if reduced[-1] != len()
+
+#def mostly_naive(string):
+#    forget, bucket_out = compute_t1(string)
+#    SA = []
+#    for i in bucket_out:
+#        SA.extend(bucket_out[i])
+#    SA = [string[SA[int(i)]:] for i in SA]
+#    SA.sort()
+#    return(SA)
+
+def completely_naive(string):
+    SA = [i for i in range(len(string))]
+    SA = [string[SA[int(i)]:] for i in SA]
+    SA.sort()

```

Comment:
This runs in $O(n^2\log n)$. It isn't great, but I can comfort you by telling you that it is far better than exponential :)